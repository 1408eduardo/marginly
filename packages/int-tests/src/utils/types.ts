import { Wallet } from 'ethers';
import { FiatTokenV2_1Contract } from '../contract-api/FiatTokenV2';
import { MarginlyPoolContract } from '../contract-api/MarginlyPool';
import { NonfungiblePositionManagerContract } from '../contract-api/NonfungiblePositionManager';
import { SwapRouterContract } from '../contract-api/SwapRouter';
import { UniswapV3PoolContract } from '../contract-api/UniswapV3Pool';
import { WETH9Contract } from '../contract-api/WETH9';

export type ContractsCollection = {
  weth: WETH9Contract;
  usdc: FiatTokenV2_1Contract;
  uniswap: UniswapV3PoolContract;
  marginly: MarginlyPoolContract;
  nonFungiblePositionManager: NonfungiblePositionManagerContract;
  swapRouter: SwapRouterContract;
};
export type Suite = (contracts: ContractsCollection, treasury: Wallet, accounts: Wallet[]) => Promise<void>;
