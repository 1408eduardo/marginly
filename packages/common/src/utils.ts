import * as ethers from 'ethers';
import { BigNumber } from '@ethersproject/bignumber';

export function formatBalance(amount: BigNumber, decimals: number, assetSymbol: string): string {
  return `${amount} (${ethers.utils.formatUnits(amount, decimals)} ${assetSymbol.toUpperCase()})`;
}

export const sleep = (ms: number) => {
  return new Promise((resolve) => setTimeout(resolve, ms));
};

export const waitForTx = async (provider: ethers.providers.Provider, hash: string) => {
  // eslint-disable-next-line no-constant-condition
  while (true) {
    const receipt = await provider.getTransactionReceipt(hash);
    if (!receipt) {
      await sleep(3000);
    } else {
      if (!receipt.status) {
        throw new Error('Transaction failed');
      }
      break;
    }
  }
};
