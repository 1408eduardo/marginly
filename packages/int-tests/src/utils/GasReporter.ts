import { ContractReceipt, ContractTransaction } from 'ethers';
import { logger } from './logger';
import { promises as fsPromises, existsSync } from 'fs';
import { join } from 'path';

type GasUsage = {
  max: number;
  min: number;
  avg: number;
  count: number;
};

export class GasReporter {
  private gasUsageStatistics: { [key: string]: GasUsage } = {};
  private gasUsage: { [key: string]: [number, number][] } = {};

  constructor(private suiteName: string) {}

  public async saveGasUsage(
    txName: string,
    x: ContractReceipt | Promise<ContractReceipt> | ContractTransaction | Promise<ContractTransaction>
  ) {
    const resolved = await x;
    let txReceipt: ContractReceipt;
    if ('wait' in resolved) {
      txReceipt = await resolved.wait();
    } else if ('gasUsed' in resolved) {
      txReceipt = resolved;
    } else {
      throw Error('Uknown argument');
    }

    const gasUsed = txReceipt.gasUsed.toNumber();
    const blockNumber = txReceipt.blockNumber;

    logger.debug(`â›½ Gas used: ${txName}    ${gasUsed}`);

    const existedStatistic = this.gasUsageStatistics[txName];
    if (existedStatistic) {
      existedStatistic.max = existedStatistic.max < gasUsed ? gasUsed : existedStatistic.max;
      existedStatistic.min = existedStatistic.min > gasUsed ? gasUsed : existedStatistic.min;
      existedStatistic.avg = Math.floor(
        (existedStatistic.avg * existedStatistic.count + gasUsed) / (existedStatistic.count + 1)
      );
      existedStatistic.count++;
    } else {
      this.gasUsageStatistics[txName] = {
        max: gasUsed,
        min: gasUsed,
        avg: gasUsed,
        count: 1,
      };
    }

    if (!this.gasUsage[txName]) {
      this.gasUsage[txName] = [[blockNumber, gasUsed]];
    } else {
      this.gasUsage[txName].push([blockNumber, gasUsed]);
    }

    return txReceipt;
  }

  public reportToConsole() {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const gasUsageStatistics = this.gasUsageStatistics;
    if (Object.keys(gasUsageStatistics).length > 0) {
      setTimeout(function () {
        console.log('Gas usage statistics');
        console.table(gasUsageStatistics);
      }, 10);
    }
  }

  async saveToFile() {
    const dir = join(__dirname, '../../', '__gas-usage__');

    if (!existsSync(dir)) {
      await fsPromises.mkdir(dir, { recursive: true });
    }

    for (const txName of Object.keys(this.gasUsage)) {
      const data = [`blockNumber,gasUsed\n`];
      for (const [blockNumber, gasUsed] of this.gasUsage[txName]) {
        data.push(`${blockNumber},${gasUsed}\n`);
      }
      const filename = `${this.suiteName}.${txName}.gas.csv`;

      await fsPromises.writeFile(join(dir, filename), data, {
        flag: 'w',
      });
    }
  }
}
