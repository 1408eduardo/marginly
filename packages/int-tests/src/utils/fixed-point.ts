import { BigNumber } from 'ethers';
import bn from 'bignumber.js';

export const FP96 = {
  one: BigNumber.from(2n ** 96n),
};

export const FP48 = {
  one: BigNumber.from(2n ** 48n),
};

// 365.25 * 24 * 60 * 60 * 2 ** 96
export const secondsInYearX96 = BigNumber.from('2500250661360148260042022567123353600');

export function pow(self: BigNumber, exponent: number): BigNumber {
  let result = FP96.one;
  while (exponent > 0) {
    if ((exponent & 1) == 1) {
      result = result.mul(self).div(FP96.one);
    }
    self = self.mul(self).div(FP96.one);
    exponent = exponent >> 1;
  }

  return result;
}

export function powTaylor(self: BigNumber, exponent: number): BigNumber {
  const x = self.sub(FP96.one);
  if (x >= FP96.one) {
    throw new Error(`x can't be greater than FP.one, series diverges`);
  }

  let resultX96 = FP96.one;
  let multiplier: BigNumber;
  let term = FP96.one;

  const steps = exponent < 3 ? exponent : 3;
  for (let i = 0; i != steps; ++i) {
    multiplier = BigNumber.from(exponent - i)
      .mul(x)
      .div(BigNumber.from(i + 1));
    term = term.mul(multiplier).div(FP96.one);
    resultX96 = resultX96.add(term);
  }

  return resultX96;
}

export function toHumanString(fp96Value: BigNumber): string {
  return bn(fp96Value.toString()).div(FP96.one.toString()).toString();
}

export function fp48ToHumanString(fp10Value: BigNumber): string {
  return bn(fp10Value.toString()).div(FP48.one.toString()).toString();
}
