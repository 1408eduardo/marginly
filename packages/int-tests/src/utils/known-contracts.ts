import { BaseContract, BigNumberish, providers, Signer } from 'ethers';
import WETH9 from '../contract-api/WETH9';
import FiatTokenV2_1 from '../contract-api/FiatTokenV2';
import UniswapV3Factory from '../contract-api/UniswapV3Factory';
import UniswapV3Pool from '../contract-api/UniswapV3Pool';
import SwapRouter from '../contract-api/SwapRouter';
import NonfungiblePositionManager from '../contract-api/NonfungiblePositionManager';

export const wethContract = (signerOrProvider?: Signer | providers.Provider) =>
  WETH9.connect(`0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2`, signerOrProvider);

export const usdcContract = (signerOrProvider?: Signer | providers.Provider) =>
  FiatTokenV2_1.connect(`0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48`, signerOrProvider);

export const uniswapFactoryContract = (signerOrProvider?: Signer | providers.Provider) =>
  UniswapV3Factory.connect(`0x1F98431c8aD98523631AE4a59f267346ea31F984`, signerOrProvider);

export const uniswapPoolContract = UniswapV3Pool.connect;

export const swapRouterContract = (signerOrProvider?: Signer | providers.Provider) =>
  SwapRouter.connect('0xE592427A0AEce92De3Edee1F18E0157C05861564', signerOrProvider);

export const nonFungiblePositionManagerContract = (signerOrProvider?: Signer | providers.Provider) =>
  NonfungiblePositionManager.connect('0xC36442b4a4522E871399CD717aBDD847Ab11FE88', signerOrProvider);

// Uniswap V3: Positions NFT 0xC36442b4a4522E871399CD717aBDD847Ab11FE88
