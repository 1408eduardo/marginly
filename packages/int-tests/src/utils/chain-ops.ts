import { logger } from '../utils/logger';
import { SECS_PER_BLOCK } from './const';
import { Web3Provider } from '@ethersproject/providers';
import { MarginlyPoolContract } from '../contract-api/MarginlyPool';
import { BigNumber, ethers, ContractReceipt } from 'ethers';

export async function waitBlocks(blocks: number): Promise<void> {
  logger.info(`Waiting for ${blocks} blocks`);
  return await new Promise((rs) => setTimeout(rs, blocks * SECS_PER_BLOCK * 1000.0));
}

export class Web3ProviderDecorator {
  readonly provider: Web3Provider;

  constructor(provider: Web3Provider) {
    this.provider = provider;
  }

  mineAtTimestamp(timestampSeconds: number): Promise<any> {
    return this.provider.send('evm_mine', [timestampSeconds]);
  }

  async getLastBlockTimestamp(): Promise<number> {
    return (await this.provider.getBlock(this.provider._lastBlockNumber)).timestamp;
  }
}

export type SwapEvent = {
  amount0: BigNumber;
  amount1: BigNumber;
  sqrtPriceX96: BigNumber;
  liquidity: BigNumber;
  tick: number;
};

export function decodeSwapEvent(txReceipt: ContractReceipt, uniswapAddress: string): SwapEvent {
  const swapEvent = txReceipt.events!.find((e) => e.address == uniswapAddress);
  const swapEventTypes = ['int256', 'int256', 'uint160', 'uint128', 'int24'];
  const result = ethers.utils.defaultAbiCoder.decode(swapEventTypes, swapEvent!.data);

  return {
    amount0: BigNumber.from(result[0]),
    amount1: BigNumber.from(result[1]),
    sqrtPriceX96: BigNumber.from(result[2]),
    liquidity: BigNumber.from(result[3]),
    tick: result[4],
  };
}

export async function getLongSortKeyX48(
  marginlyPool: MarginlyPoolContract,
  accountAddress: string
): Promise<BigNumber> {
  const position = await marginlyPool.positions(accountAddress);
  const index = BigNumber.from(position.heapPosition).sub(1);
  logger.debug(`  heap position is ${position.heapPosition}`);
  const [, leverage] = await marginlyPool.getLongHeapPosition(index);
  return BigNumber.from(leverage.key);
}

export async function getShortSortKeyX48(
  marginlyPool: MarginlyPoolContract,
  accountAddress: string
): Promise<BigNumber> {
  const position = await marginlyPool.positions(accountAddress);
  const index = BigNumber.from(position.heapPosition).sub(1);
  logger.debug(`  heap position is ${position.heapPosition}`);
  const [, leverage] = await marginlyPool.getShortHeapPosition(index);
  return BigNumber.from(leverage.key);
}
