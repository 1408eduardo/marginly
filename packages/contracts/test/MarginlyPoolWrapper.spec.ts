import { loadFixture } from '@nomicfoundation/hardhat-network-helpers';
import { ethers } from 'hardhat';
import { createMarginlyPoolWithWrapper } from './shared/fixtures';
import { FP96, PositionType } from './shared/utils';
import { expect } from 'chai';
import snapshotGasCost from '@uniswap/snapshot-gas-cost';

describe('MarginlyPoolWrapper long', () => {
  it('success', async () => {
    const { marginlyPoolWrapper, marginlyPool } = await loadFixture(createMarginlyPoolWithWrapper);
    const [_, signer, lender] = await ethers.getSigners();

    const depositBaseAmount = 1000n;
    const depositQuoteAmount = 5000n;
    const longAmount = 2000n;

    await marginlyPool.connect(lender).depositBase(depositBaseAmount);
    await marginlyPool.connect(lender).depositQuote(depositQuoteAmount);

    await marginlyPoolWrapper.connect(signer).long(depositBaseAmount, longAmount);

    const baseCollCoeff = await marginlyPool.baseCollateralCoeff();

    const position = await marginlyPool.positions(signer.address);
    expect(position._type).to.be.equal(PositionType.Long);
    expect(position.discountedBaseAmount.mul(baseCollCoeff)).to.be.equal((depositBaseAmount + longAmount) * FP96.one);

    // Must not exist
    const wrapperContractPosition = await marginlyPool.positions(marginlyPoolWrapper.address);
    expect(wrapperContractPosition._type).to.be.equal(PositionType.Uninitialized);
    expect(wrapperContractPosition.discountedBaseAmount).to.be.equal(0);
    expect(wrapperContractPosition.discountedQuoteAmount).to.be.equal(0);

  });

  it('fail', async () => {
    const { marginlyPoolWrapper, marginlyPool } = await loadFixture(createMarginlyPoolWithWrapper);
    const [_, signer, lender] = await ethers.getSigners();

    const depositBaseAmount = 1000n;
    const depositQuoteAmount = 5000n;
    const longAmount = 1n;

    await marginlyPool.connect(lender).depositBase(depositBaseAmount);
    await marginlyPool.connect(lender).depositQuote(depositQuoteAmount);

    await expect(marginlyPoolWrapper.connect(signer).long(depositBaseAmount, longAmount)).to.be.revertedWith('MA');

    const position = await marginlyPool.positions(signer.address);
    expect(position._type).to.be.equal(PositionType.Uninitialized);
    expect(position.discountedBaseAmount).to.be.equal(0);
    expect(position.discountedQuoteAmount).to.be.equal(0);

    // Must not exist
    const wrapperContractPosition = await marginlyPool.positions(marginlyPoolWrapper.address);
    expect(wrapperContractPosition._type).to.be.equal(PositionType.Uninitialized);
    expect(wrapperContractPosition.discountedBaseAmount).to.be.equal(0);
    expect(wrapperContractPosition.discountedQuoteAmount).to.be.equal(0);

  });

  it('gas cost snapshot', async () => {
    const { marginlyPoolWrapper, marginlyPool } = await loadFixture(createMarginlyPoolWithWrapper);
    const [_, signer, lender] = await ethers.getSigners();

    const depositBaseAmount = 1000n;
    const depositQuoteAmount = 5000n;
    const longAmount = 2000n;

    await marginlyPool.connect(lender).depositBase(depositBaseAmount);
    await marginlyPool.connect(lender).depositQuote(depositQuoteAmount);
    await snapshotGasCost(await marginlyPoolWrapper.connect(signer).long(depositBaseAmount, longAmount));
  });
});

describe('MarginlyPoolWrapper short', () => {
  it('success', async () => {
    const { marginlyPoolWrapper, marginlyPool } = await loadFixture(createMarginlyPoolWithWrapper);
    const [_, signer, lender] = await ethers.getSigners();

    const depositBaseAmount = 5000n;
    const depositQuoteAmount = 1000n;
    const shortAmount = 2000n;

    await marginlyPool.connect(lender).depositBase(depositBaseAmount);
    await marginlyPool.connect(lender).depositQuote(depositQuoteAmount);

    await marginlyPoolWrapper.connect(signer).short(depositQuoteAmount, shortAmount);

    const quoteCollCoeff = await marginlyPool.quoteCollateralCoeff();
    const price = (await marginlyPool.getBasePrice()).inner;

    const position = await marginlyPool.positions(signer.address);
    expect(position._type).to.be.equal(PositionType.Short);
    const expected = price.mul(shortAmount).add(depositQuoteAmount * FP96.one).div(FP96.one);
    expect(position.discountedQuoteAmount.mul(quoteCollCoeff).div(FP96.one)).to.be.equal(expected);

    // Must not exist
    const wrapperContractPosition = await marginlyPool.positions(marginlyPoolWrapper.address);
    expect(wrapperContractPosition._type).to.be.equal(PositionType.Uninitialized);
    expect(wrapperContractPosition.discountedBaseAmount).to.be.equal(0);
    expect(wrapperContractPosition.discountedQuoteAmount).to.be.equal(0);
  });

  it('fail', async () => {
    const { marginlyPoolWrapper, marginlyPool } = await loadFixture(createMarginlyPoolWithWrapper);
    const [_, signer, lender] = await ethers.getSigners();

    const depositBaseAmount = 1000n;
    const depositQuoteAmount = 1000n;
    const shortAmount = 1n;

    await marginlyPool.connect(lender).depositBase(depositBaseAmount);
    await marginlyPool.connect(lender).depositQuote(depositQuoteAmount);

    await expect(marginlyPoolWrapper.connect(signer).short(depositQuoteAmount, shortAmount)).to.be.revertedWith('MA');

    const position = await marginlyPool.positions(signer.address);
    expect(position._type).to.be.equal(PositionType.Uninitialized);
    expect(position.discountedBaseAmount).to.be.equal(0);
    expect(position.discountedQuoteAmount).to.be.equal(0);

    // Must not exist
    const wrapperContractPosition = await marginlyPool.positions(marginlyPoolWrapper.address);
    expect(wrapperContractPosition._type).to.be.equal(PositionType.Uninitialized);
    expect(wrapperContractPosition.discountedBaseAmount).to.be.equal(0);
    expect(wrapperContractPosition.discountedQuoteAmount).to.be.equal(0);

  });

  it('gas cost snapshot', async () => {
    const { marginlyPoolWrapper, marginlyPool } = await loadFixture(createMarginlyPoolWithWrapper);
    const [_, signer, lender] = await ethers.getSigners();

    const depositBaseAmount = 5000n;
    const depositQuoteAmount = 1000n;
    const shortAmount = 2000n;

    await marginlyPool.connect(lender).depositBase(depositBaseAmount);
    await marginlyPool.connect(lender).depositQuote(depositQuoteAmount);
    await snapshotGasCost(await marginlyPoolWrapper.connect(signer).short(depositQuoteAmount, shortAmount));
  });
});