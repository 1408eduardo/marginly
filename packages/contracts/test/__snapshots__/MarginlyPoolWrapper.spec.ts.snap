// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarginlyPoolWrapper long gas cost snapshot 1`] = `
Object {
  "calldataByteLength": 100,
  "gasUsed": 439718,
}
`;

exports[`MarginlyPoolWrapper short gas cost snapshot 1`] = `
Object {
  "calldataByteLength": 100,
  "gasUsed": 436640,
}
`;
